### ISSUE TITLE
vendor prefixes do not have priority

### ISSUE THREAD

**Issue by tostiheld** [phpactor/phpactor#1746]
Problem:

- The first `@return` is always accepted regardless if there is a better typed prefixed (e.g. `@psalm-`) alternative
- The `?T` doesn't seem to be recognised as a template parameter, see comments.

---

I'm struggling a bit with autocompleting generics, specifically in combination with a Doctrine ServiceEntityRepository. Psalm does seem to understand everything I'm doing, however phpactor does not. Here's an example:

```
/**
 * @template T of object
 * @template-extends ServiceEntityRepository<T>
 */
abstract class AbstractRepository extends ServiceEntityRepository
{
    /** @return T */
    public function get(Stringable $id)
    {
        $entity = $this->find($id);
        if (null === $entity) {
            throw new \Exception();
        }

        return $entity;
    }
}
/**
 * @extends AbstractRepository<SomeEntity>
 */
class SomeEntityRepository extends AbstractRepository
{
    public function __construct(ManagerRegistry $registry)
    {
        parent::__construct($registry, SomeEntity::class);
    }
}
```

When I try to see the return type of SomeEntityRepository::get(), it gives me T instead of SomeEntity . Like I said, Psalm understands I mean SomeEntity . Am I doing something wrong here? Is this beyond the (current) capabilities of phpactor? (edited)

**Comment by dantleech** [comment]
Phpactor's generic support is not currently complete but I think this should work. I'll add this to the test suite shortly and check.

**Comment by dantleech** [comment]
![image](https://user-images.githubusercontent.com/530801/178159873-b5faea70-ebc1-444f-96e7-b0372c3242b1.png)

this seems to work as far as resolving the variable type?

**Comment by dantleech** [comment]
Also seems fine when calling `get`:

![image](https://user-images.githubusercontent.com/530801/178159971-4a8603cc-871f-4579-8bb6-9307080ea205.png)

**Comment by dantleech** [comment]
hmm, but I guess more to do with the ServiceEntityRepository

ok, by replacing

```
     * @return object|null The entity instance or NULL if the entity can not be found.
     * @psalm-return ?T
```

with

```
     * @psalm-return T
```

it "works": 

![image](https://user-images.githubusercontent.com/530801/178160199-bb166662-7948-4744-9426-34d71f711b76.png)

So two problems:

- Prioritising prefixed annotations before non-prefixed ones
- Template parameter not recognised if nullable (?)

**Comment by dantleech** [comment]
Fixed the null issue in #1762 deciding which docblock tag to use when presented with prefixed tags and "normal" tags is a more challenging problem.

**Comment by dantleech** [comment]
As the first issue is resolved, I've created a new issue for the remaining one: https://github.com/phpactor/phpactor/issues/1780
